opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java

```diff
@@ -982,89 +997,84 @@ private static AnalyzerFactory findForStream(InputStream in,
             }
             content = Arrays.copyOf(content, len);
         }
-
-        AnalyzerFactory fac;
-
-        // First, do precise-magic Matcher matching
-        for (FileAnalyzerFactory.Matcher matcher : matchers) {
-            if (matcher.isPreciseMagic()) {
-                fac = matcher.isMagic(content, in);
-                if (fac != null) {
-                    if (LOGGER.isLoggable(Level.FINEST)) {
-                        LOGGER.log(Level.FINEST,
-                            "''{0}'': chosen by precise magic: {1}",
-                                new Object[]{file, fac.getClass().getSimpleName()});
-                    }
-                    return fac;
-                }
-            }
-        }
-
-        // Next, look for magic strings
-        String opening = readOpening(in, content);
-        fac = findMagicString(opening, file);
-        if (fac != null) {
-            return fac;
+        var finalContent = Arrays.copyOf(content, len);
+        try {
+            return findFactoryUsingMatcher(finalContent, in, file, true)
+                    .or(() -> findUsingMagicString(finalContent, in, file))
+                    .or(() -> findFactoryUsingMatcher(finalContent, in, file, false))
+                    .orElse(null);
+        } catch (WrapperIOException ex) {
+            throw ex.getParentIOException();
         }
+    }
 
-        // Last, do imprecise-magic Matcher matching
-        for (FileAnalyzerFactory.Matcher matcher : matchers) {
-            if (!matcher.isPreciseMagic()) {
-                fac = matcher.isMagic(content, in);
-                if (fac != null) {
-                    if (LOGGER.isLoggable(Level.FINEST)) {
-                        LOGGER.log(Level.FINEST,
-                            "''{0}'': chosen by imprecise magic: {1}",
-                            new Object[]{file, fac.getClass().getSimpleName()});
+    private static Optional<AnalyzerFactory> findFactoryUsingMatcher(byte[] content, InputStream in,
+                                                                 String file,
+                                                              boolean isPrecise) {
+        var optionalFactory = matchers.stream()
```

Comment:
I wonder if this could benefit from `paralleStream()`

Reply:
Parallel Stream have additional overhead of creating the records, splitting collection etc, so it is recommend to  user sequential streams for most of the usecases.
Parallel Stream have better performance if the collection size is huge or if we are doing many computation for each element.
Considering these factors I think sequential stream is a better fit here.
https://www.baeldung.com/java-when-to-use-parallel-stream
https://stackoverflow.com/questions/23170832/java-8s-streams-why-parallel-stream-is-slower